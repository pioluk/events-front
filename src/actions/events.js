import {
  FETCH_EVENTS_REQUEST,
  FETCH_EVENTS_SUCCESS,
  FETCH_EVENTS_FAILURE
} from '../constants/events'

export function fetchEventsRequest () {
  return {
    type: FETCH_EVENTS_REQUEST
  }
}

export function fetchEventsSuccess (events) {
  return {
    type: FETCH_EVENTS_SUCCESS,
    events
  }
}

export function fetchEventsFailure (error) {
  return {
    type: FETCH_EVENTS_FAILURE,
    error
  }
}
