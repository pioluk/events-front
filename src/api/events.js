import { apiGet } from './utils'
import { API_URL } from './config'

export const getEvent = id => apiGet('event/' + id)

export const getEvents = () => apiGet('event')

export const createEvent = event => {
  const user = JSON.parse(localStorage.getItem('user')) || {}

  const formData = new FormData()

  const [ startDate, _1 ] = event.dateStart.toJSON().split('T')
  const [ _2, startTime ] = event.timeStart.toJSON().split('T')

  const [ endDate, _3 ] = event.dateEnd.toJSON().split('T')
  const [ _4, endTime ] = event.timeEnd.toJSON().split('T')

  formData.append('UserId', user.id)
  formData.append('title', event.title)
  formData.append('description', event.description)
  formData.append('dateStart', startDate + 'T' + startTime)
  formData.append('dateEnd', endDate + 'T' + endTime)
  formData.append('color', event.color)
  formData.append('image', event.image)
  formData.append('emails', event.emails)
  formData.append('phones', event.phones)
  formData.append('websites', event.websites)
  formData.append('location.name', event.location.name)
  formData.append('location.lat', event.location.lat)
  formData.append('location.lng', event.location.lng)
  formData.append('location.placeId', event.location.placeId)

  return fetch(`${API_URL}/event`, {
    method: 'POST',
    body: formData
  })
  .then(response => {
    if (response.status >= 200 && response.status < 300) {
      return response
    }

    const error = new Error(response.statusText)
    error.response = response
    throw error
  })
  .then(x => x.json())
}
