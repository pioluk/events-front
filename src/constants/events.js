export const FETCH_EVENTS_REQUEST = 'FETCH_EVENTS_REQUEST'
export const FETCH_EVENTS_SUCCESS = 'FETCH_EVENTS_SUCCESS'
export const FETCH_EVENTS_FAILURE = 'FETCH_EVENTS_FAILURE'
export const ADD_EVENT = 'ADD_EVENT'
export const ADD_EVENT_SUCCESS = 'ADD_EVENT_SUCCESS'
export const ADD_EVENT_FAILURE = 'ADD_EVENT_FAILURE'
