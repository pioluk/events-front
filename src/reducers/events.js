/* @flow */

import {
  FETCH_EVENTS_REQUEST,
  FETCH_EVENTS_SUCCESS,
  FETCH_EVENTS_FAILURE
} from '../constants/events'

type Event = any

type EventsState = {
  isLoading: boolean,
  events: Array<Event>,
  error: ?Error
}

const initialState: EventsState = {
  isLoading: false,
  events: [],
  error: null
}

export default function eventsReducer (state: EventsState = initialState, action: any) {
  switch (action.type) {
    case FETCH_EVENTS_REQUEST:
      return { ...state, isLoading: true }

    case FETCH_EVENTS_SUCCESS:
      return {
        ...state,
        isLoading: false,
        events: [...state.events, ...action.events]
      }

    case FETCH_EVENTS_FAILURE:
      return { ...state, isLoading: false, events: [], error: action.error }

    default:
      return state
  }
}
