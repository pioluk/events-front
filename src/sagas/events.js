import { call, take, put } from 'redux-saga/effects'
import { FETCH_EVENTS_REQUEST, FETCH_EVENT_DETAILS_REQUEST, ADD_EVENT } from '../constants/events'
import * as actionCreators from '../actions/events'
import { getEvent, getEvents, createEvent } from '../api/events'

export function* getEventsFlow () {
  while (true) {
    yield take(FETCH_EVENTS_REQUEST)
    try {
      const events = yield call(getEvents)
      if (events) {
        yield put(actionCreators.fetchEventsSuccess(events))
      }
    }
    catch (err) {
      yield put(actionCreators.fetchEventsFailure(err))
    }
  }
}

export function* getEventDetailsFlow () {
  while (true) {
    const { eventId } = yield take(FETCH_EVENT_DETAILS_REQUEST)
    try {
      const eventDetails = yield call(getEvent, eventId)
      if (eventDetails) {
        yield put(actionCreators.fetchEventDetailsSuccess(eventDetails))
      }
    }
    catch (err) {
      yield put(actionCreators.fetchEventDetailsFailure(err))
    }
  }
}

export function* createEventFlow () {
  while (true) {
    const { event } = yield take(ADD_EVENT)
    try {
      yield call(createEvent, event)
      yield put(actionCreators.addEventSuccess())
    }
    catch (err) {
      yield put(actionCreators.addEventFailure(err))
    }
  }
}
