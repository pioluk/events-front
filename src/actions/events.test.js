// global describe,it,expect

import * as actions from './events'

describe('events actions', () => {
  it('should create an action to fetch events', () => {
    expect(actions.fetchEventsRequest()).toMatchSnapshot()
  })

  it('should create an action to send fetched events', () => {
    const events = [
      {
        "id": 1,
        "UserId": 1,
        "title": "nihil voluptas rerum",
        "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
        "dateStart": "2016-05-23T17:38:11.010Z",
        "dateEnd": "2016-12-06T02:03:56.063Z",
        "color": "a5dfd8",
        "createdAt": "2016-10-26T07:06:39.152Z",
        "updatedAt": "2016-10-26T07:06:39.152Z"
      },
      {
        "id": 2,
        "UserId": 1,
        "title": "voluptatem eos nostrum",
        "description": "Sequi sit minima quae totam dolores suscipit. Vero exercitationem repellendus velit. Earum ullam unde explicabo earum. Corporis architecto ab quisquam consectetur. Odio odit aliquid ex sint consequatur similique. Aut et occaecati saepe rem cupiditate et veritatis aut. Similique molestias eveniet libero labore est laboriosam. Molestiae reiciendis error neque dignissimos.",
        "dateStart": "2016-07-04T18:33:59.969Z",
        "dateEnd": "2017-08-27T03:37:53.566Z",
        "color": "33ed3f",
        "createdAt": "2016-10-26T07:06:40.243Z",
        "updatedAt": "2016-10-26T07:06:40.243Z"
      }
    ]
    expect(actions.fetchEventsSuccess(events)).toMatchSnapshot()
  })

  it('should create an action to indicate event fetching error', () => {
    const error = new Error('Error fetching error')
    expect(actions.fetchEventsFailure(error)).toMatchSnapshot()
  })

  it('should create an action to request creation of a new event', () => {
    const event = {
      "id": 1,
      "UserId": 1,
      "title": "nihil voluptas rerum",
      "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
      "dateStart": "2016-05-23T17:38:11.010Z",
      "dateEnd": "2016-12-06T02:03:56.063Z",
      "color": "a5dfd8",
      "createdAt": "2016-10-26T07:06:39.152Z",
      "updatedAt": "2016-10-26T07:06:39.152Z"
    }
    expect(actions.addEvent(event)).toMatchSnapshot()
  })

  it('should create an action to add a new event', () => {
    const event = {
      "id": 1,
      "UserId": 1,
      "title": "nihil voluptas rerum",
      "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
      "dateStart": "2016-05-23T17:38:11.010Z",
      "dateEnd": "2016-12-06T02:03:56.063Z",
      "color": "a5dfd8",
      "createdAt": "2016-10-26T07:06:39.152Z",
      "updatedAt": "2016-10-26T07:06:39.152Z"
    }
    expect(actions.addEventSuccess(event)).toMatchSnapshot()
  })

  it('should create an action to indicate an event creation error', () => {
    const error = new Error('Event creation error.')
    expect(actions.addEventFailure(error)).toMatchSnapshot()
  })

  it('should create an action to fetch event\'s details', () => {
    const eventId = 1
    expect(actions.fetchEventDetailsReqeust(eventId)).toMatchSnapshot()
  })

  it('should create an action to pass event details', () => {
    const eventDetails = {
      "id": 1,
      "UserId": 1,
      "title": "nihil voluptas rerum",
      "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
      "dateStart": "2016-05-23T17:38:11.010Z",
      "dateEnd": "2016-12-06T02:03:56.063Z",
      "color": "a5dfd8",
      "createdAt": "2016-10-26T07:06:39.152Z",
      "updatedAt": "2016-10-26T07:06:39.152Z"
    }
    expect(actions.fetchEventDetailsSuccess(eventDetails)).toMatchSnapshot()
  })

  it('should create an action to indicate event\'s details fetching error', () => {
    const error = new Error('Event details fetching failed.')
    expect(actions.fetchEventDetailsFailure(error)).toMatchSnapshot()
  })
})
