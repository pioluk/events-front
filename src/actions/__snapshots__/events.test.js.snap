exports[`events actions should create an action to add a new event 1`] = `
Object {
  "event": Object {
    "UserId": 1,
    "color": "a5dfd8",
    "createdAt": "2016-10-26T07:06:39.152Z",
    "dateEnd": "2016-12-06T02:03:56.063Z",
    "dateStart": "2016-05-23T17:38:11.010Z",
    "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
    "id": 1,
    "title": "nihil voluptas rerum",
    "updatedAt": "2016-10-26T07:06:39.152Z",
  },
  "type": "ADD_EVENT_SUCCESS",
}
`;

exports[`events actions should create an action to fetch event's details 1`] = `
Object {
  "eventId": 1,
  "type": "FETCH_EVENT_DETAILS_REQUEST",
}
`;

exports[`events actions should create an action to fetch events 1`] = `
Object {
  "page": 1,
  "type": "FETCH_EVENTS_REQUEST",
}
`;

exports[`events actions should create an action to increase page 1`] = `
Object {
  "type": "INCREASE_PAGE",
}
`;

exports[`events actions should create an action to indicate an event creation error 1`] = `
Object {
  "error": [Error: Event creation error.],
  "type": "ADD_EVENT_FAILURE",
}
`;

exports[`events actions should create an action to indicate event fetching error 1`] = `
Object {
  "error": [Error: Error fetching error],
  "type": "FETCH_EVENTS_FAILURE",
}
`;

exports[`events actions should create an action to indicate event's details fetching error 1`] = `
Object {
  "error": [Error: Event details fetching failed.],
  "type": "FETCH_EVENT_DETAILS_FAILURE",
}
`;

exports[`events actions should create an action to pass event details 1`] = `
Object {
  "eventDetails": Object {
    "UserId": 1,
    "color": "a5dfd8",
    "createdAt": "2016-10-26T07:06:39.152Z",
    "dateEnd": "2016-12-06T02:03:56.063Z",
    "dateStart": "2016-05-23T17:38:11.010Z",
    "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
    "id": 1,
    "title": "nihil voluptas rerum",
    "updatedAt": "2016-10-26T07:06:39.152Z",
  },
  "type": "FETCH_EVENT_DETAILS_SUCCESS",
}
`;

exports[`events actions should create an action to request creation of a new event 1`] = `
Object {
  "event": Object {
    "UserId": 1,
    "color": "a5dfd8",
    "createdAt": "2016-10-26T07:06:39.152Z",
    "dateEnd": "2016-12-06T02:03:56.063Z",
    "dateStart": "2016-05-23T17:38:11.010Z",
    "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
    "id": 1,
    "title": "nihil voluptas rerum",
    "updatedAt": "2016-10-26T07:06:39.152Z",
  },
  "type": "ADD_EVENT",
}
`;

exports[`events actions should create an action to send fetched events 1`] = `
Object {
  "count": Array [
    Object {
      "UserId": 1,
      "color": "a5dfd8",
      "createdAt": "2016-10-26T07:06:39.152Z",
      "dateEnd": "2016-12-06T02:03:56.063Z",
      "dateStart": "2016-05-23T17:38:11.010Z",
      "description": "Ut ut ipsam. Hic voluptas illo ex voluptatibus dolores aut sed beatae vel. Quo fuga cum animi est voluptatem. Animi numquam magni. Ut labore deleniti tenetur nulla magnam maxime enim. Molestiae omnis harum id sequi accusantium tempora. Tempore est dolorem nesciunt voluptatem quaerat qui dolorum non. Praesentium perferendis ut.",
      "id": 1,
      "title": "nihil voluptas rerum",
      "updatedAt": "2016-10-26T07:06:39.152Z",
    },
    Object {
      "UserId": 1,
      "color": "33ed3f",
      "createdAt": "2016-10-26T07:06:40.243Z",
      "dateEnd": "2017-08-27T03:37:53.566Z",
      "dateStart": "2016-07-04T18:33:59.969Z",
      "description": "Sequi sit minima quae totam dolores suscipit. Vero exercitationem repellendus velit. Earum ullam unde explicabo earum. Corporis architecto ab quisquam consectetur. Odio odit aliquid ex sint consequatur similique. Aut et occaecati saepe rem cupiditate et veritatis aut. Similique molestias eveniet libero labore est laboriosam. Molestiae reiciendis error neque dignissimos.",
      "id": 2,
      "title": "voluptatem eos nostrum",
      "updatedAt": "2016-10-26T07:06:40.243Z",
    },
  ],
  "events": undefined,
  "type": "FETCH_EVENTS_SUCCESS",
}
`;
